File: 프로그래머스/2/42747. H－Index/H－Index.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<int> citations) {
    int answer = 0;
    sort(citations.begin(), citations.end());
    
    for (int i = 0; i < max((int)citations.size() + 1, *(citations.rbegin()) + 1); i++) {
        if (citations.end() - lower_bound(citations.begin(), citations.end(), i) >= i) {
            answer = i;
        }
    }
    return answer;
}

File: 프로그래머스/2/42747. H－Index/H－Index.java
import java.util.*;

class Solution {
    public int solution(int[] citations) {
        int answer = 0;
        
        Arrays.sort(citations);
        
        int left = 0, right = citations[citations.length - 1];
        while (left <= right) {
            int mid = (left + right) / 2;
            
            int count = 0;
            for (int i = 0; i < citations.length; i++) {
                if (citations[i] >= mid) {
                    count = citations.length - i;
                    break;
                }
            }
            
            if (count >= mid) {
                answer = mid;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        
        return answer;
    }
}